src/commonMain/kotlin/teksturepako/pakku/cli/cmd/Diff.kt

```diff
@@ -29,128 +33,338 @@ class Diff : CliktCommand()
+    {
+        val updatedProjects = mutableListOf<String>()
+        val verboseUpdatedFiles = mutableMapOf<String, String>()
+
+        for (newProject in allNewProjects)
+        {
+            /** We only care about projects, which previously also existed **/
+            val oldProject = allOldProjects.find { it.isAlmostTheSameAs(newProject) } ?: continue
 
+            /** If the project files are not identical, means that files have changed **/
+            if (oldProject.files == newProject.files) continue
+
+            /** On multi-loader modpacks, the mod might have been added for another loader, causing the
+             * files key to have changed without an actual update.
+             **/
+            val oldFileHash = oldProject.files.firstOrNull()?.hashes?.get("sha1")
+            val newFileHash = newProject.files.firstOrNull()?.hashes?.get("sha1")
+            if (oldFileHash == newFileHash) continue
+
+            if (verboseOpt)
+            {
+                val oldFileName = oldProject.files.firstOrNull()?.fileName
+                val newFileName = newProject.files.firstOrNull()?.fileName
+                val fileNamesNotNullOrEmpty = !oldFileName.isNullOrEmpty() && !newFileName.isNullOrEmpty()
+
+                if (fileNamesNotNullOrEmpty) verboseUpdatedFiles["$oldFileName"] = "$newFileName"
+            }
+            else
+            {
+                updatedProjects.add("${newProject.name.values.firstOrNull()}")
+            }
+        }
         if (verboseOpt)
         {
             val maxOldFileLength = verboseUpdatedFiles.keys.maxOfOrNull { it.length } ?: 0
             for ((oldFileName, newFileName) in verboseUpdatedFiles)
             {
-                updated.add("${oldFileName.padEnd(maxOldFileLength)} -> $newFileName")
+                updatedProjects.add("${oldFileName.padEnd(maxOldFileLength)} -> $newFileName")
             }
         }
-        updated.forEach { terminal.info("! $it") }
+        return updatedProjects
+    }
 
+    private fun printDiffChangesToTerminal(
+        addedMCVersions: Set<String>,
+        removedMCVersions: Set<String>,
+        addedModLoaders: Set<String>,
+        removedModLoaders: Set<String>,
+        updatedModLoaders: Set<String>,
+        addedProjects: Set<String>,
+        removedProjects: Set<String>,
+        updatedProjects: List<String>
+    )
+    {
+        addedMCVersions.forEach { terminal.success("+ $it") }
+        removedMCVersions.forEach { terminal.danger("- $it") }
+        addedModLoaders.forEach { terminal.success("+ $it") }
+        removedModLoaders.forEach { terminal.danger("- $it") }
+        updatedModLoaders.forEach { terminal.info("! $it") }
+        addedProjects.forEach { terminal.success("+ $it") }
+        removedProjects.forEach { terminal.danger("- $it") }
+        updatedProjects.forEach { terminal.info("! $it") }
+    }
```

Comment:
Instead of doing it like this, you could do it like this:

```kt
private fun printDiffChangesToTerminal(added: Set<String>, removed: Set<String>, updated: Set<String> = setOf())
{
    added.forEach { terminal.success("+ $it") }
    removed.forEach { terminal.danger("- $it") }
    updated.forEach { terminal.success("! $it") }
}
```